### ISSUE TITLE
[docs]: DIAGRAM - Implement "Operator nesting" in "Diagram Canvas"

### ISSUE THREAD

**Issue by BioPhoton** [BioPhoton/Rx-Marble-Design-System#18]
!Notice! => values are wrong bit not the visualisation!

In some situations, we need to display nested operators.
At the moment this is not solved. Below you see different versions of the `share` operator. 
This is a very good example of the need for nested operators because:
- `Subject` is used in `multicast`
- `multicast` and `refCount` is used in `share`

https://docs.google.com/presentation/d/1LlVmTEcoc1LRV00Tp-rDp4kopI-NobC9S2w-w5pC_JA/edit#slide=id.g4c256b0d88_0_79

![Subjects](https://user-images.githubusercontent.com/10064416/62827417-b7092e00-bbce-11e9-929e-0d862870af45.png)
---
![multicast](https://user-images.githubusercontent.com/10064416/62827419-bb354b80-bbce-11e9-9b94-cbe85290ea36.png)
![refCount](https://user-images.githubusercontent.com/10064416/62827420-bcff0f00-bbce-11e9-9ae6-17f155b08f0b.png)
---
![RxJS Multicast Operators-67](https://user-images.githubusercontent.com/10064416/62827398-99d45f80-bbce-11e9-8ab2-d8623db8b5c9.png)
![RxJS Multicast Operators-68](https://user-images.githubusercontent.com/10064416/62827399-9a6cf600-bbce-11e9-9e95-5b3a38b9deae.png)
![RxJS Multicast Operators-69](https://user-images.githubusercontent.com/10064416/62827400-9a6cf600-bbce-11e9-8a50-dffb51b8e9ea.png)
![RxJS Multicast Operators-70](https://user-images.githubusercontent.com/10064416/62827401-9a6cf600-bbce-11e9-9745-8659b100d1a2.png)
![RxJS Multicast Operators-71](https://user-images.githubusercontent.com/10064416/62827402-9b058c80-bbce-11e9-8f49-93d9a095e989.png)
![RxJS Multicast Operators-72](https://user-images.githubusercontent.com/10064416/62827403-9b058c80-bbce-11e9-94e9-bdb18cd002fa.png)
![RxJS Multicast Operators-73](https://user-images.githubusercontent.com/10064416/62827404-9b9e2300-bbce-11e9-808d-530411272685.png)
![RxJS Multicast Operators-74](https://user-images.githubusercontent.com/10064416/62827405-9b9e2300-bbce-11e9-8d65-b4ee48cac0b7.png)
![RxJS Multicast Operators-75](https://user-images.githubusercontent.com/10064416/62827406-9b9e2300-bbce-11e9-9a86-656421f566c1.png)
![RxJS Multicast Operators-76](https://user-images.githubusercontent.com/10064416/62827407-9c36b980-bbce-11e9-9fc8-9ba8bed34be8.png)
![RxJS Multicast Operators-77](https://user-images.githubusercontent.com/10064416/62827408-9ccf5000-bbce-11e9-8c66-739be4c42b68.png)
![RxJS Multicast Operators-78](https://user-images.githubusercontent.com/10064416/62827409-9ccf5000-bbce-11e9-88d7-eff30cab0ffd.png)
![RxJS Multicast Operators-79](https://user-images.githubusercontent.com/10064416/62827410-9ccf5000-bbce-11e9-840d-d56263c2f58a.png)

**Comment by BioPhoton** [comment]
@cedricss maybe you have some input :)

**Comment by kyleryanbanks** [comment]
So I'm trying to run through and just make sure I understand what I'm looking at first before offering any thoughts.

Is the 0-1-2-3 in the s1$ stream inside the share examples just a copy paste from the multicast example and we're just focusing on the overall UI rather than specific values?  It seems like that should be 1-2-3-4 and I just want to make sure I'm following correctly.

**Comment by BioPhoton** [comment]
@kyleryanbanks you are right! 
The numbers are wrong! Sry for this confusion.
Anyway thanks a lot for your time! 🙏

**Comment by kyleryanbanks** [comment]
I think you might be able to use bold lines and 1 block indentation to show nesting.

I would avoid shading since that breaks down as soon as you run out of shades of grey for each level of nesting.

Only had a notebook and a pen on me atm, but I'll try to link to a picture of what I was thinking.  

It might be nice to have some way to show the nesting in the flow description, too. I think the bold limits you to one level.

If you don't want to show every single nested level (eg: share > multicast > subject ) then I think you need some way to "black box" a higher level operator so people know there's logic they don't see, but can expand, jump to, or peek it's diagram easily.

I think you lose a little bit of real estate with the indention, but I think you get more clarity of nesting.  I think it's okay to have them extend all the way to the right side to show internal operators that remain open at the end if the diagram.

![1566157413146243799636557970177](https://user-images.githubusercontent.com/36783391/63229509-993b5a80-c1c6-11e9-9865-0585a95b00eb.jpg)

(Everything should probably shift over a few blocks to fix the subscribe and connect, but I couldn't maintain all of that in my head manually, lol.)

I think this would make building diagrams easier since nested operators should almost just be drag and drop off the lower level diagrams, too.

**Comment by kyleryanbanks** [comment]
Saw the note on Twitter regarding the time shift, might be able to use the time block to cover any growth from the nests and just show that it was still just 1 frame?

I think that maybe only becomes an issue when something nested wouldn't happen until after the first value is emitted, though, right?

Otherwise you're just shifting the start point?

**Comment by BioPhoton** [comment]
Yes, you are right with the starting point. But the frames are normally starting at the left border of the operator...

Here some new versions:

![Marble Design System Nesting Operators](https://user-images.githubusercontent.com/10064416/63230739-3622e580-c211-11e9-913f-5b5ebfaacc6b.png)
![Marble Design System Nesting Operators (1)](https://user-images.githubusercontent.com/10064416/63230751-594d9500-c211-11e9-9062-2e715abc23b0.png)
![Marble Design System Nesting Operators (2)](https://user-images.githubusercontent.com/10064416/63230742-3ae79980-c211-11e9-9903-7442f03e8710.png)
![Marble Design System Nesting Operators (3)](https://user-images.githubusercontent.com/10064416/63230743-3c18c680-c211-11e9-9712-24963c96b30d.png)
![Marble Design System Nesting Operators (4)](https://user-images.githubusercontent.com/10064416/63230744-3d49f380-c211-11e9-91bf-497336f8efb8.png)

Happy for any feedback @kyleryanbanks :)

**Comment by kyleryanbanks** [comment]
So I like simple 2 and 3.  I think it's clearer where the nested functions end on 3, though.

For 2 it's hard to tell if the refcount is that entire bottom chunk with a weird operator stream or if that's the bottom line.

Maybe the flow descriptions for any nested operators can just be links until a more official black box solution gets made?

**Comment by kyleryanbanks** [comment]
It feels reasonable to only show 1 layer of nested detail (assuming the lower levels have diagrams somewhere else in the docs.)
